src/modules/fancyzones/lib/JsonHelpers.cpp

```diff
@@ -0,0 +1,951 @@
+
+    // From Settings.cs
+    constexpr int c_focusModelId = 0xFFFF;
+    constexpr int c_rowsModelId = 0xFFFE;
+    constexpr int c_columnsModelId = 0xFFFD;
+    constexpr int c_gridModelId = 0xFFFC;
+    constexpr int c_priorityGridModelId = 0xFFFB;
+    constexpr int c_blankCustomModelId = 0xFFFA;
+
+}
+
+namespace JSONHelpers
+{
+    json::JsonArray NumVecToJsonArray(const std::vector<int>& vec)
+    {
+        json::JsonArray arr;
+        for (const auto& val : vec)
+        {
+            arr.Append(json::JsonValue::CreateNumberValue(val));
+        }
+
+        return arr;
+    }
+
+    std::vector<int> JsonArrayToNumVec(const json::JsonArray& arr)
+    {
+        std::vector<int> vec;
+        for (const auto& val : arr)
+        {
+            vec.emplace_back(static_cast<int>(val.GetNumber()));
+        }
+
+        return vec;
+    }
+
+    ZoneSetLayoutType TypeFromLayoutId(int layoutID)
+    {
+        switch (layoutID)
+        {
+        case c_focusModelId:
+            return ZoneSetLayoutType::Focus;
+        case c_columnsModelId:
+            return ZoneSetLayoutType::Columns;
+        case c_rowsModelId:
+            return ZoneSetLayoutType::Rows;
+        case c_gridModelId:
+            return ZoneSetLayoutType::Grid;
+        case c_priorityGridModelId:
+            return ZoneSetLayoutType::PriorityGrid;
+        default:
+            return ZoneSetLayoutType::Custom;
+        }
+    }
+
+    std::wstring TypeToString(ZoneSetLayoutType type)
+    {
+        switch (type)
+        {
+        case ZoneSetLayoutType::Focus:
+            return L"focus";
+        case ZoneSetLayoutType::Columns:
+            return L"columns";
+        case ZoneSetLayoutType::Rows:
+            return L"rows";
+        case ZoneSetLayoutType::Grid:
+            return L"grid";
+        case ZoneSetLayoutType::PriorityGrid:
+            return L"priority-grid";
+        case ZoneSetLayoutType::Custom:
+            return L"custom";
+        default:
+            return L"TypeToString_ERROR";
+        }
+    }
+
+    ZoneSetLayoutType TypeFromString(const std::wstring& typeStr)
+    {
+        if (typeStr.compare(L"focus") == 0)
+        {
+            return JSONHelpers::ZoneSetLayoutType::Focus;
+        }
+        else if (typeStr.compare(L"columns") == 0)
+        {
+            return JSONHelpers::ZoneSetLayoutType::Columns;
+        }
+        else if (typeStr.compare(L"rows") == 0)
+        {
+            return JSONHelpers::ZoneSetLayoutType::Rows;
+        }
+        else if (typeStr.compare(L"grid") == 0)
+        {
+            return JSONHelpers::ZoneSetLayoutType::Grid;
+        }
+        else if (typeStr.compare(L"priority-grid") == 0)
+        {
+            return JSONHelpers::ZoneSetLayoutType::PriorityGrid;
+        }
+        else
+        { //Custom
+            return JSONHelpers::ZoneSetLayoutType::Custom;
+        }
+    }
+
+    FancyZonesData& FancyZonesDataInstance()
+    {
+        static FancyZonesData instance;
```

Comment:
First part (marking every instance method of FancyZonesDataInstance as const where it's possible) is done.

Reply:
Alright, now we need to make sure there're no const/non-const and non-const/const concurrent invocations.

Reply:
I would prefer to have a mutex and be sure we don't crash.
Later we can revisit the implementation and in case remove the mutex where is not strictly needed, but for now the top priority is to not crash and also to make sure we don't have data corruptions.

Reply:
I've opened https://github.com/microsoft/PowerToys/issues/1246